### ISSUE TITLE
OF 1.3 switches

### ISSUE THREAD

**Issue by Neby1992** [floodlight/floodlight#633]
When switch send some information about packet on port, there is parsing error.
15:19:32.390 ERROR [n.f.c.i.OFSwitchHandshakeHandler:nioEventLoopGroup-3-4] OFBadRequestErrorMsgVer13(xid=117, code=BAD_PORT, data=[unparsed: 04 0d 00 73 00 00 00 75 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 02 00 00 00 00 00 00 00 00 01 80 c2 00 00 0e 5c 8a 38 c9 89 7f 88 cc 02 07 04 5c 8a 38 c9 89 55 04]) from switch OFSwitch DPID[00:01:5c:8a:38:c9:89:55] in state net.floodlightcontroller.core.internal.OFSwitchHandshakeHandler$MasterState@78f0a824
15:19:32.397 WARN [n.f.c.i.Controller:nioEventLoopGroup-3-4] Unhandled OF Message: OFBadRequestErrorMsgVer13(xid=117, code=BAD_PORT, data=[unparsed: 04 0d 00 73 00 00 00 75 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 02 00 00 00 00 00 00 00 00 01 80 c2 00 00 0e 5c 8a 38 c9 89 7f 88 cc 02 07 04 5c 8a 38 c9 89 55 04]) from OFSwitch DPID[00:01:5c:8a:38:c9:89:55]
15:19:32.400 ERROR [n.f.c.i.OFSwitchHandshakeHandler:nioEventLoopGroup-3-4] OFBadRequestErrorMsgVer13(xid=118, code=BAD_PORT, data=[unparsed: 04 0d 00 73 00 00 00 76 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 04 00 00 00 00 00 00 00 00 01 80 c2 00 00 0e 5c 8a 38 c9 89 81 88 cc 02 07 04 5c 8a 38 c9 89 55 04]) from switch OFSwitch DPID[00:01:5c:8a:38:c9:89:55] in state net.floodlightcontroller.core.internal.OFSwitchHandshakeHandler$MasterState@78f0a824
15:19:32.404 WARN [n.f.c.i.Controller:nioEventLoopGroup-3-4] Unhandled OF Message: OFBadRequestErrorMsgVer13(xid=118, code=BAD_PORT, data=[unparsed: 04 0d 00 73 00 00 00 76 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 04 00 00 00 00 00 00 00 00 01 80 c2 00 00 0e 5c 8a 38 c9 89 81 88 cc 02 07 04 5c 8a 38 c9 89 55 04]) from OFSwitch DPID[00:01:5c:8a:38:c9:89:55]
15:19:32.461 ERROR [n.f.c.i.OFSwitchHandshakeHandler:nioEventLoopGroup-3-4] OFBadRequestErrorMsgVer13(xid=128, code=BAD_PORT, data=[unparsed: 04 0d 00 7b 00 00 00 80 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 02 00 00 00 00 00 00 00 00 ff ff ff ff ff ff 5c 8a 38 c9 89 7f 89 42 20 00 06 04 00 02 00 00 02 07]) from switch OFSwitch DPID[00:01:5c:8a:38:c9:89:55] in state net.floodlightcontroller.core.internal.OFSwitchHandshakeHandler$MasterState@78f0a824
15:19:32.464 WARN [n.f.c.i.Controller:nioEventLoopGroup-3-4] Unhandled OF Message: OFBadRequestErrorMsgVer13(xid=128, code=BAD_PORT, data=[unparsed: 04 0d 00 7b 00 00 00 80 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 02 00 00 00 00 00 00 00 00 ff ff ff ff ff ff 5c 8a 38 c9 89 7f 89 42 20 00 06 04 00 02 00 00 02 07]) from OFSwitch DPID[00:01:5c:8a:38:c9:89:55]
15:19:32.467 ERROR [n.f.c.i.OFSwitchHandshakeHandler:nioEventLoopGroup-3-4] OFBadRequestErrorMsgVer13(xid=129, code=BAD_PORT, data=[unparsed: 04 0d 00 7b 00 00 00 81 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 04 00 00 00 00 00 00 00 00 ff ff ff ff ff ff 5c 8a 38 c9 89 81 89 42 20 00 06 04 00 02 00 00 02 07]) from switch OFSwitch DPID[00:01:5c:8a:38:c9:89:55] in state net.floodlightcontroller.core.internal.OFSwitchHandshakeHandler$MasterState@78f0a824
15:19:32.468 WARN [n.f.c.i.Controller:nioEventLoopGroup-3-4] Unhandled OF Message: OFBadRequestErrorMsgVer13(xid=129, code=BAD_PORT, data=[unparsed: 04 0d 00 7b 00 00 00 81 ff ff ff ff ff ff ff ff 00 10 00 00 00 00 00 00 00 00 00 10 00 00 00 04 00 00 00 00 00 00 00 00 ff ff ff ff ff ff 5c 8a 38 c9 89 81 89 42 20 00 06 04 00 02 00 00 02 07]) from OFSwitch DPID[00:01:5c:8a:38:c9:89:55]

**Comment by rizard** [comment]
@Neby1992 what switch are you using? HP comware switches are known to do this. It's a firmware issue on the switch AFAIK.

**Comment by Neby1992** [comment]
I have been using HP 5130 with comware 7. I had caught communication between switch and controller. The switch send port 4 but controller did not resolve it. I tried to find section where controller check port name but I did not find it.

**Comment by rizard** [comment]
@Neby1992 The parsing error is on the HP side. It's not a parsing error actually. The OFBadRequest message is generated by the switch when the controller tries to insert flows using certain ports. What I've found is that HP advertises all its ports X, Y, Z, etc. Only the set of ports that are 40G (or possibly higher) are usable by the controller, from my experience. Any flow referencing a 10G port results in the errors you have provided. The errors might be able to be safely ignored if you can determine that the flows are actually being inserted.

Long story short -- I've run into this before myself and determined it was a problem with the switch firmware. Maybe you have an update available?

**Comment by Neby1992** [comment]
@rizard I have been using the newest firmware which HP released. When I tried HP VAN SDN controller everthing works. I will try to write to the HP. And I will try tu push flow for sigle port. 
Another problem is RouterOS which send information about ports which is avaliable, but floodlight GUI do not shows ports.
Thanks for response rizard.

**Comment by rizard** [comment]
@Neby1992 Interesting. Can you compare the flow-mod/packet-out messages Floodlight sends to those the HP VAN controller sends? Maybe the ports are not the root cause, and maybe there's something else the switch doesn't like.

**Comment by Neby1992** [comment]
So I tested and FloodLight sent Barrier request and HP send reply.
But I do not know what does it mean.

Frame 62: 386 bytes on wire (3088 bits), 386 bytes captured (3088 bits) on interface 0
Ethernet II, Src: HewlettP_c9:89:68 (5c:8a:38:c9:89:68), Dst: HewlettP_05:d5:af (14:58:d0:05:d5:af)
Internet Protocol Version 4, Src: 192.168.103.247, Dst: 192.168.103.248
Transmission Control Protocol, Src Port: 6925 (6925), Dst Port: 6633 (6633), Seq: 4573, Ack: 809, Len: 320
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_BARRIER_REPLY (21)
    Length: 8
    Transaction ID: 4294967289
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_BARRIER_REPLY (21)
    Length: 8
    Transaction ID: 4294967288
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_ERROR (1)
    Length: 76
    Transaction ID: 18
    Type: OFPET_BAD_REQUEST (1)
    Code: OFPBRC_BAD_PORT (11)
    Data
        Version: 1.3 (0x04)
        Type: OFPT_PACKET_OUT (13)
        Length: 115
        Transaction ID: 18
        Body: ffffffffffffffff00100000000000000000001000000014...
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_ERROR (1)
    Length: 76
    Transaction ID: 19
    Type: OFPET_BAD_REQUEST (1)
    Code: OFPBRC_BAD_PORT (11)
    Data
        Version: 1.3 (0x04)
        Type: OFPT_PACKET_OUT (13)
        Length: 115
        Transaction ID: 19
        Body: ffffffffffffffff00100000000000000000001000000003...
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_ERROR (1)
    Length: 76
    Transaction ID: 20
    Type: OFPET_BAD_REQUEST (1)
    Code: OFPBRC_BAD_PORT (11)
    Data
OpenFlow 1.3
    Version: 1.3 (0x04)
    Type: OFPT_ERROR (1)
    Length: 76
    Transaction ID: 21
    Type: OFPET_BAD_REQUEST (1)
    Code: OFPBRC_BAD_PORT (11)
    Data

**Comment by rizard** [comment]
The barrier request and replies aren't the problem. They're the packet-out
messages Floodlight is sending for link discovery. Floodlight just uses the
ports it has learned when it handshakes with the switch. The controller
should be able to reference any of the ports in packet-outs and flow-mods,
but it's not able to, which is the problem

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by Neby1992** [comment]
Is possible that probles is incoming port? Because controller send OF_ANNY and in OF1.3 speciffication is valid port or OFPP_CONTROLLER

**Comment by rizard** [comment]
It could be. I'm almost certain it's the output port though. You can test
with the 40G uplinks and see that they don't generate the errors. At least
that was the case for me.

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by Neby1992** [comment]
OK I will try, but I would like to change the port. Do you know where this code is?

**Comment by rizard** [comment]
What port do you want to change?

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by Neby1992** [comment]
I would like to change the input port in the message which is send automaticaly by Floodlight.

**Comment by Neby1992** [comment]
I found it in LinkDiscoveryManager.java. Should I create pull request?

**Comment by rizard** [comment]
Does that fix the problem?

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by rizard** [comment]
Specifically, what did you change, and what is the result after the change?

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by Neby1992** [comment]
It looks good and works, but I will be tested it and I will let you know. And thank you for help.
I change pob.setInPort(OFPort.ANY); to pob.setInPort(OFPort.CONTROLLER);

**Comment by rizard** [comment]
Haha that's interesting. The HP has been the only one that's given that
problem. Thanks for looking into it for us! Yes, please issue a pull
request if you don't mind. I'll acknowledge you as a contributor. Thanks so
much!

Ryan Izard
PhD Candidate, Research/Teaching Assistant
306B Fluor Daniel Building
ECE Department, Clemson University
Clemson, SC 29634
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Comment by Neby1992** [comment]
I made pull request, but building was unsucessfull, but not in my code :D

**Comment by rizard** [comment]
@Neby1992 looks like a periodic issue with openjdk and the unit tests. It's not your fault, so I'll go ahead and merge #636. If you'd like to look into the failed unit test, feel free. But, I'll warn you that it'll be tough to debug. They all pass on my machines (Mac + Linux) using both JDKs.

**Comment by raghdaaljaff** [comment]
Hello can somebody help me please Iam trying to implement the flowvisor in my floodlight controller and when I wrote this command sudo -u flowvisor fvconfig generate /etc/flowvisor /config.json
The error is : sudo :unknown user:flowvisor
sudo :unable to initialize policy plugin.
